File: libs/lib_ui/src/main/java/com/ycr/lib/ui/pullrefresh/smart/SmartRefreshView.java
package com.ycr.lib.ui.pullrefresh.smart;

import android.content.Context;
import android.util.AttributeSet;

import com.ycr.lib.ui.pullrefresh.IPullListener;
import com.ycr.lib.ui.pullrefresh.IRefreshListener;
import com.ycr.lib.ui.pullrefresh.IRefreshView;
import com.scwang.smartrefresh.layout.SmartRefreshLayout;
import com.scwang.smartrefresh.layout.api.RefreshLayout;
import com.scwang.smartrefresh.layout.constant.RefreshState;
import com.scwang.smartrefresh.layout.listener.OnLoadMoreListener;
import com.scwang.smartrefresh.layout.listener.OnRefreshListener;

/**
 * Created by ye on 2018/2/4.
 */

public class SmartRefreshView extends SmartRefreshLayout implements IRefreshView {
    private SmartRefreshOnMultiPurposeListener smartRefreshOnMultiPurposeListener;

    public SmartRefreshView(Context context) {
        super(context);
        init();
    }

    public SmartRefreshView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public SmartRefreshView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    public void init() {
        setPullLoadEnabled(false);
        setPullRefreshEnabled(true);
        smartRefreshOnMultiPurposeListener = new SmartRefreshOnMultiPurposeListener();
        super.setOnMultiPurposeListener(smartRefreshOnMultiPurposeListener);
    }

    @Override
    public void autoHeaderRefreshing(int wait) {
//        resetStatus();
        super.autoRefresh(wait);
    }

    @Override
    public boolean isPullRefreshing() {
        return getState()== RefreshState.Refreshing;
    }

    @Override
    public boolean isPullupLoading() {
        return getState()== RefreshState.Loading;
    }

    @Override
    public void setPullRefreshEnabled(boolean pullRefreshEnabled) {
        super.setEnableRefresh(pullRefreshEnabled);
    }

    @Override
    public void setPullLoadEnabled(boolean pullLoadEnabled) {
        super.setEnableLoadMore(pullLoadEnabled);
    }

    @Override
    public boolean isPullRefreshEnabled() {
        return mEnableRefresh;
    }

    @Override
    public boolean isPullLoadEnabled() {
        return mEnableLoadMore;
    }

    @Override
    public void onPullDownRefreshComplete(boolean isSuccess) {
        super.finishRefresh(isSuccess);
    }

    @Override
    public void onPullUpRefreshComplete(boolean isSuccess) {
        super.finishLoadMore(isSuccess);
    }

    @Override
    public void setLastUpdatedLabel(CharSequence label) {

    }

    @Override
    public long getSmoothScrollDuration() {
        return 400;
    }

    @Override
    public void setHasMoreData(boolean hasMoreData, boolean showHasMoreData) {
        super.setNoMoreData(!hasMoreData);//设置之后，将不会再触发加载事件
    }


    @Override
    public void setRefreshListener(final IRefreshListener refreshListener) {
        if (refreshListener == null) {
            return;
        }
        super.setOnRefreshListener(new OnRefreshListener() {
            @Override
            public void onRefresh(RefreshLayout refreshlayout) {
                refreshListener.onPullrefresh(SmartRefreshView.this);
            }
        });
        super.setOnLoadMoreListener(new OnLoadMoreListener() {
            @Override
            public void onLoadMore(RefreshLayout refreshlayout) {
                refreshListener.onLoadMore(SmartRefreshView.this);
            }
        });
    }

    @Override
    public void setPullListener(IPullListener pullListener) {
        if (smartRefreshOnMultiPurposeListener != null) {
            smartRefreshOnMultiPurposeListener.setPullListener(pullListener);
        }
    }

}

File: libs/lib_ui/src/main/java/com/ycr/lib/ui/pullrefresh/smart/SmartRefreshView.kt
package com.ycr.lib.ui.pullrefresh.smart

import android.content.Context
import android.support.annotation.Nullable
import android.util.AttributeSet
import com.ycr.lib.ui.pullrefresh.*

import com.scwang.smartrefresh.layout.SmartRefreshLayout
import com.scwang.smartrefresh.layout.constant.RefreshState

/**
 * 实现了统一刷新接口的拉动刷新控件
 * Created by yuchengren on 2018/06/14.
 */

class SmartRefreshView @JvmOverloads constructor(context: Context, @Nullable attrs: AttributeSet? = null, defStyleAttr: Int = 0) :
        SmartRefreshLayout(context, attrs, defStyleAttr) , IRefreshView {

    companion object{
        const val ANIMATION_DURATION = 300 //动画时长
    }

    private val smartRefreshOnMultiPurposeListener: SmartRefreshOnMultiPurposeListener =
            SmartRefreshOnMultiPurposeListener()

    override val isPullDownRefreshing: Boolean
        get() = state == RefreshState.Refreshing

    override val isPullUpLoadingMore: Boolean
        get() = state == RefreshState.Loading

    override var isPullDownRefreshEnabled: Boolean
        get() = mEnableRefresh
        set(value) {
            super.setEnableRefresh(value)
        }

    override var isPullUpLoadMoreEnabled: Boolean
        get() = mEnableLoadMore
        set(value) {
            super.setEnableLoadMore(value)
        }

    override var noMoreData: Boolean
        get() = mFooterNoMoreData
        set(value) {
            super.setNoMoreData(value)
        }

    override var isFooterVisibleWhenNoMoreData: Boolean
        get() = mEnableFooterFollowWhenNoMoreData
        set(value) {
            super.setEnableFooterFollowWhenNoMoreData(value)
        }

    init{
        super.setOnMultiPurposeListener(smartRefreshOnMultiPurposeListener)
    }

    override fun autoPullDownRefresh(delayed: Int) {
        super.autoRefresh(delayed)
    }

    override fun finishPullDownRefresh(isSuccess: Boolean) {
        super.finishRefresh(isSuccess)
    }

    override fun finishPullDownRefresh(isSuccess: Boolean, noMoreData: Boolean) {
        val passTime = System.currentTimeMillis() - mLastOpenTime
        super.finishRefresh(if (isSuccess) Math.min(Math.max(0, ANIMATION_DURATION - passTime.toInt()), ANIMATION_DURATION) else 0, isSuccess, noMoreData)//保证加载动画有300毫秒的时间
    }

    override fun finishPullUpLoadMore(isSuccess: Boolean) {
        super.finishLoadMore(isSuccess)
    }

    override fun finishPullUpLoadMore(isSuccess: Boolean, noMoreData: Boolean) {
        val passTime = System.currentTimeMillis() - mLastOpenTime
        super.finishLoadMore(if (isSuccess) Math.min(Math.max(0, ANIMATION_DURATION - passTime.toInt()), ANIMATION_DURATION) else 0, isSuccess, noMoreData)
    }

    override fun setOnPullDownRefreshListener(onPullDownRefreshListener: OnPullDownRefreshListener) {
        super.setOnRefreshListener { onPullDownRefreshListener.onPullDownRefresh(this) }
    }

    override fun setOnPullUpLoadMoreListener(onPullUpLoadMoreListener: OnPullUpLoadMoreListener) {
        super.setOnLoadMoreListener { onPullUpLoadMoreListener.onPullUpLoadMore(this) }
    }

    override fun setOnPullStateChangedListener(pullListener: OnPullStateChangedListener) {
        smartRefreshOnMultiPurposeListener.setPullListener(pullListener)
    }
}
